MetaMask/metamask-extension
# Import a token after a user receives it from a confirmed transaction

## Explanation
@cloudonshore noticed that transactions that DApps create for users to sign on MetaMask that result in token transfers to users does not result in those tokens being automatically added to the user's app state.

This is especially problematic when users use DApps to swap for new tokens that they don't already have, as this will essentially result in the funds "disappearing" to the user until they add the asset.

This may result in many erroneous support requests about "incorrect" or "missing" assets or balances as well as a poor overall user experience.

This PR addresses this issue by creating a mechanism by which:
1) When pending transactions are confirmed, MetaMask will loop through the transaction receipt's transaction logs and identify all erc20 transfers
2) Filter the transfers by ones where the recipient of the transfer is the sender of the transaction (currently, this is assumed to be the user's account. note: this isn't necessarily true when paymasters are used - but i don't think MetaMask does this for now).
3) For each unique `log.address` (asset contract address), the app will save the token info in the same mechanism as "Import Token" while firing off an analytics event indicating the token was added (@ todo: this analytics event should probably be called something different?)

## Manual Testing Steps
- Find a DApp that allows the user to swap from one asset to another.
- Find a new asset that isn't already added to the your state.
- Create a transaction to exchange from some known asset to a new asset using a DApp with MetaMask as the transaction signer.
- Approve transaction on MetaMask.
- Wait for transaction to confirm.
- Check that the new asset is added to the user's assets list.

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
